### PR TITLE
DA Improve Packer Build and Get MySQL Working with PHP 7 or Later

### PR DESCRIPTION
**Pull request by Dala0**
Changes Packer build to get a working version of PHP 7 using RPM and Yum. Also adds working Nginx and PHP-FPM configuration files.

Tested on a machine running Mac OS 10.11.

Changes to the PHP code that make it work with version seven or later. The changes are largely to the code that interacts with MySQL. This was needed because the old MySQL functions where removed in PHP version seven.

Tested on a virtual machine running Linux in Amazon Web Services.

@fredex42

### CODE REVIEW

**Comment by Dala0** [comment]
[Get encodings endpoint running on up-to-date OS and PHP](https://trello.com/c/mTYyYMEp/2663-get-encodings-endpoint-running-on-up-to-date-os-and-php)

File: packer/endpointserver.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      #- rm -rf memcached-3.0.4
+      #- echo ------------------------------------
+      #- echo Setting up php setttings
+      #- echo ------------------------------------
+
+      #- sudo touch /usr/local/lib/php.ini
+      #- sudo chown ec2-user /usr/local/lib/php.ini
+      #- cat > /usr/local/lib/php.ini << EOF
+      #- extension=memcached.so
+      #- error_log=/var/log/php/error.log
```
Comment (fredex42) [inline packer/endpointserver.yaml]: don't we need this? where do error logs go?

File: packer/endpointserver.yaml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      #- cd ..
+      #- rm -rf php-7.2.2*
+
+      - sudo rpm -Uvh http://rpms.famillecollet.com/enterprise/remi-release-7.rpm
+
+      - echo ------------------------------------
+      - echo Installing PHP
+      - echo ------------------------------------
+      
+      - sudo yum -y --enablerepo=remi,remi-php72 install php php-common php-mysql php-fpm php-pecl-memcache php-pecl-memcached
```
Comment (fredex42) [inline packer/endpointserver.yaml]: how trustworthy is this rpm source do you think? how up-to-date is it kept?

File: packer/endpointserver.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  - type: shell
+    inline:
+      #needed to ensure that awscli actually installs
+      #https://bugzilla.redhat.com/show_bug.cgi?id=1417304
+      - sudo yum-config-manager --enable rhui-REGION-rhel-server-optional
+      - sudo yum -y update
+      - sudo rpm -ivh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
+
+      - sudo yum -y install epel-release
+      - sudo yum -y groupinstall 'Development Tools'
```
Comment (fredex42) [inline packer/endpointserver.yaml]: don't think we need this if not building, but working on that right now

File: packer/endpointserver.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      #- sudo chown ec2-user /usr/local/lib/php.ini
+      #- cat > /usr/local/lib/php.ini << EOF
+      #- extension=memcached.so
+      #- error_log=/var/log/php/error.log
+      #- EOF
+      #- sudo chown root /usr/local/lib/php.ini
+      #install composer
+      - sudo curl -s "http://getcomposer.org/installer" | php
+      - echo Setting up selinux security context
+      - sudo mkdir -p /var/www
```
Comment (fredex42) [inline packer/endpointserver.yaml]: need to blank out default page

**Review by fredex42 — COMMENTED** [review]
couple of notes, am working on getting a docker build as well for circleci so hold off making changes for now

File: packer/endpointserver.yaml (hunk 1 above)
Comment (Dala0) [reply packer/endpointserver.yaml]: I expect so. I think I commented it by mistake because it was near lines which needed disabling.

Comment (Dala0) [reply packer/endpointserver.yaml]: Actually, that php.ini file is no longer used by the current system, so it does not matter.

File: packer/endpointserver.yaml (hunk 2 above)
Comment (Dala0) [reply packer/endpointserver.yaml]: It looks pretty trustworthy to me. It was last updated on 22/6/2018.

File: packer/endpointserver.yaml (hunk 3 above)
Comment (Dala0) [reply packer/endpointserver.yaml]: Possibly not, but it may break something if removed, so I say keep it.

File: packer/endpointserver.yaml (hunk 1 above)
Comment (fredex42) [reply packer/endpointserver.yaml]: ok let's work out where the logs go though, we need to make sure that they get picked up by filebeat

Comment (Dala0) [reply packer/endpointserver.yaml]: The current system uses the file at /etc/php.ini

Comment (Dala0) [reply packer/endpointserver.yaml]: I had a look in this file. The error_log value is not set. Maybe we should write a value for it into the /etc/php.ini file?

@fredex42

File: circle.yml.old.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+machine:
```
Comment (fredex42) [inline circle.yml.old.yml]: can remote this file now

File: endpoint/common.php [hunk 6]
```diff
@@ -133,30 +133,29 @@ function find_content($config){
 		$data = $mc->get($_SERVER['REQUEST_URI']);
 		if($data){
 			if(array_key_exists('status',$data) && $data['status']=='notfound') return null;
-		#	print "Cache hit!\n";
+			#print "Cache hit!\n";
```
Comment (fredex42) [inline endpoint/common.php]: can remove these lines or change them to debug statements

File: endpoint/common.php [hunk 7]
```diff
@@ -133,30 +133,29 @@ function find_content($config){
...
 		}
 	}
-	
+		
 	$num_servers = count($config['dbhost']);
 
 	$n = 0;
 	$dbh=false;
 	while(!$dbh){
 		//print "Trying to connect to database at ".$config['dbhost'][$n]." (attempt $n)\n";
```
Comment (fredex42) [inline endpoint/common.php]: likewise can remove or change to debug statement

File: endpoint/tests/_files/apache_config.conf [hunk 8]
```diff
@@ -1,7 +1,7 @@
 Listen 8080
```
Comment (fredex42) [inline endpoint/tests/_files/apache_config.conf]: is this file needed any more? think we're using nginx for the integration test now

File: endpoint/tests/phpunit.xml [hunk 9]
```diff
@@ -15,9 +15,9 @@
         <var name="DB_DSN" value="mysql:dbname=circle_test;host=127.0.0.1;port=3306" />
         <!--<var name="DB_DSN" value="sqlite::memory"/>-->
```
Comment (fredex42) [inline endpoint/tests/phpunit.xml]: remove commented lline, put a comment in here to remind to keep user id and password in-sync with `.circleci/config.yml`

File: endpoint/tests/test_bootstrap.php [hunk 10]
```diff
@@ -2,7 +2,8 @@
 $dbname = $GLOBALS['DB_DBNAME'];
 try {
     $pdo = new PDO($GLOBALS['DB_DSN'], $GLOBALS['DB_USER'], $GLOBALS['DB_PASSWD'] );
-} catch(PDOException $e){
+} catch(PDOException $e) {
+    print "Got exception $e when setting up database";
```
Comment (fredex42) [inline endpoint/tests/test_bootstrap.php]: can remove this print line now

File: packer/endpointserver.yaml [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  - type: amazon-ebs
+    ami_name: Packer build for endpointserver 8
+    #get a bit more compute, we are compiling stuff
+    instance_type: t2.large
+    source_ami: ami-c90195b0
+    ssh_username: ec2-user
+  - type: docker
+    image: fedora:28  #keep this in sync with http://rpms.famillecollet.com/fedora/remi-release-28.rpm around line 51
+    commit: true
+#built 1st version at ami-7ca0cc05. need to check how to install memcached
```
Comment (fredex42) [inline packer/endpointserver.yaml]: can remove these comments, they are out of date

File: packer/endpointserver.yaml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      # needed for Docker build
+      - if [ ! -x /usr/bin/sudo ]; then yum -y install sudo yum-utils; fi
+      #needed to ensure that awscli actually installs
+      #https://bugzilla.redhat.com/show_bug.cgi?id=1417304
+      - sudo yum-config-manager --enable rhui-REGION-rhel-server-optional
+      - sudo yum -y update
+      - if [ ! -f /etc/fedora-release ]; then sudo rpm -ivh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm; fi
+
+      - if [ ! -f /etc/fedora-release ]; then sudo yum -y install epel-release; fi
+      # - sudo yum -y groupinstall 'Development Tools'
```
Comment (fredex42) [inline packer/endpointserver.yaml]: pls remove commented out lines

**Review by fredex42 — CHANGES_REQUESTED** [review]
great, just a bit of tidying up to do 😄

**Comment by Dala0** [comment]
Nothing controversial if your latest requested changes, so I have attempted all of them.

@fredex42

**Review by fredex42 — APPROVED** [review]
:+1 happy with this
